File: backend/radar-service/src/main/java/com/radovan/spring/services/impl/EurekaRegistrationServiceImpl.java
package com.radovan.spring.services.impl;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpEntity;
import org.springframework.http.HttpHeaders;
import org.springframework.http.MediaType;
import org.springframework.scheduling.annotation.Scheduled;
import org.springframework.stereotype.Service;
import org.springframework.web.client.RestTemplate;

import com.radovan.spring.services.EurekaRegistrationService;

import tools.jackson.databind.JsonNode;
import tools.jackson.databind.ObjectMapper;

import java.net.InetAddress;
import java.util.HashMap;
import java.util.Map;

@Service
public class EurekaRegistrationServiceImpl implements EurekaRegistrationService {

	private static final String EUREKA_SERVER_URL = "http://localhost:8761/eureka/apps";

	@Autowired
	private RestTemplate restTemplate;

	@Override
	@Scheduled(fixedRate = 30000L)
	public void registerService() {
		ObjectMapper objectMapper = new ObjectMapper();

		try {
			// Dinamičko dohvaćanje IP adrese i hostname-a
			String hostname = InetAddress.getLocalHost().getHostName();
			String ipAddr = InetAddress.getLocalHost().getHostAddress();

			String appName = "radar-service";
			String instanceId = appName + "-01";
			int port = 8082;

			// Kreiranje podataka za registraciju
			Map<String, Object> instanceData = new HashMap<>();
			instanceData.put("instanceId", instanceId);
			instanceData.put("app", appName);
			instanceData.put("hostName", hostname); // Koristi hostname
			instanceData.put("ipAddr", ipAddr); // Dinamički dohvaćena IP adresa
			instanceData.put("statusPageUrl", "http://" + ipAddr + ":" + port + "/info");
			instanceData.put("healthCheckUrl", "http://" + ipAddr + ":" + port + "/api/health");
			instanceData.put("homePageUrl", "http://" + ipAddr + ":" + port + "/");
			instanceData.put("vipAddress", appName);
			instanceData.put("secureVipAddress", appName);
			instanceData.put("leaseRenewalIntervalInSeconds", 30);
			instanceData.put("leaseExpirationDurationInSeconds", 90);

			// Dodavanje porta
			Map<String, Object> portMap = new HashMap<>();
			portMap.put("$", port);
			instanceData.put("port", portMap);

			Map<String, Object> securePortMap = new HashMap<>();
			securePortMap.put("$", 0);
			instanceData.put("securePort", securePortMap);

			// Dodavanje metadata
			Map<String, Object> metadata = new HashMap<>();
			metadata.put("management.port", port);
			instanceData.put("metadata", metadata);

			// Dodavanje dataCenterInfo
			Map<String, Object> dataCenterInfo = new HashMap<>();
			dataCenterInfo.put("@class", "com.netflix.appinfo.InstanceInfo$DefaultDataCenterInfo");
			dataCenterInfo.put("name", "MyOwn");
			instanceData.put("dataCenterInfo", dataCenterInfo);

			// Glavni JSON objekat sa "instance" ključem
			Map<String, Object> registrationData = new HashMap<>();
			registrationData.put("instance", instanceData);

			// Konverzija u JSONNode
			JsonNode jsonPayload = objectMapper.valueToTree(registrationData);

			// Postavljanje Content-Type zaglavlja
			HttpHeaders headers = new HttpHeaders();
			headers.setContentType(MediaType.APPLICATION_JSON);

			// Kreiranje HttpEntity sa JSON telom i zaglavljem
			HttpEntity<JsonNode> requestEntity = new HttpEntity<>(jsonPayload, headers);

			// Slanje POST zahteva
			String registrationUrl = EUREKA_SERVER_URL + "/" + appName;
			restTemplate.postForEntity(registrationUrl, requestEntity, JsonNode.class);

		} catch (Exception e) {
			e.printStackTrace();
			throw new RuntimeException("Failed to register service with Eureka", e);
		}
	}
}

File: backend/api-gateway/src/main/scala/com/radovan/scalatra/services/impl/EurekaRegistrationServiceImpl.scala
package com.radovan.scalatra.services.impl

import com.radovan.scalatra.config.ApacheHttpClientSync
import com.radovan.scalatra.services.EurekaRegistrationService
import jakarta.inject.{Inject, Singleton}
import org.apache.pekko.actor.{ActorSystem, Cancellable}
import org.apache.pekko.stream.Materializer

import java.net.InetAddress
import flexjson.JSONSerializer

import scala.collection.mutable
import scala.concurrent.ExecutionContext
import scala.jdk.CollectionConverters._

@Singleton
class EurekaRegistrationServiceImpl @Inject() (
                                                system: ActorSystem,
                                                mat: Materializer
                                              ) extends EurekaRegistrationService {

  implicit val ec: ExecutionContext = system.dispatcher

  private val eurekaServerUrl = "http://localhost:8761/eureka/apps"
  private val appName = "api-gateway"
  private val port = 8080

  private val scheduler: Cancellable = system.scheduler.scheduleAtFixedRate(
    initialDelay = scala.concurrent.duration.Duration.Zero,
    interval = scala.concurrent.duration.Duration(30, "seconds")
  )(() => registerService())

  override def registerService(): Unit = {

    try {
      val hostname = InetAddress.getLocalHost.getHostName
      val ipAddr = InetAddress.getLocalHost.getHostAddress
      val instanceId = "api-gateway-01"

      val instanceData = mutable.LinkedHashMap[String, Any]()

      instanceData.put("instanceId", "api-gateway-01")
      instanceData.put("app", "api-gateway")
      instanceData.put("hostName", hostname)
      instanceData.put("ipAddr", ipAddr)
      instanceData.put("statusPageUrl", s"http://$ipAddr:$port/info")
      instanceData.put("healthCheckUrl", s"http://$ipAddr:$port/api/health")
      instanceData.put("homePageUrl", s"http://$ipAddr:$port/")
      instanceData.put("vipAddress", "api-gateway")
      instanceData.put("secureVipAddress", "api-gateway")
      instanceData.put("leaseRenewalIntervalInSeconds", 30)
      instanceData.put("leaseExpirationDurationInSeconds", 90)

      val portMap = mutable.LinkedHashMap[String, Any]()
      portMap.put("$", port)
      portMap.put("@enabled", true)
      instanceData.put("port", portMap.asJava)

      val securePortMap = mutable.LinkedHashMap[String, Any]()
      securePortMap.put("$", 0)
      securePortMap.put("@enabled", false)
      instanceData.put("securePort", securePortMap.asJava)

      val metadata = mutable.LinkedHashMap[String, Any]()
      metadata.put("management.port", port)

      val dataCenterInfo = mutable.LinkedHashMap[String, Any]()
      dataCenterInfo.put("@class", "com.netflix.appinfo.InstanceInfo$DefaultDataCenterInfo")
      dataCenterInfo.put("name", "MyOwn")

      // Ubacivanje u instanceData
      instanceData.put("metadata", metadata.asJava)
      instanceData.put("dataCenterInfo", dataCenterInfo.asJava)

      val registrationData = mutable.LinkedHashMap[String, Any]()
      registrationData.put("instance", instanceData.asJava)

      val jsonString = new JSONSerializer()
        .exclude("*.class")
        .deepSerialize(registrationData.asJava)

      val headers = Map(
        "Content-Type" -> "application/json",
        "Accept" -> "application/json"
      )

      val registrationUrl = s"$eurekaServerUrl/${appName.toUpperCase}"


      val (statusCode, responseBody) = ApacheHttpClientSync.post(
        registrationUrl,
        jsonString,
        Map.empty
      )



      if (statusCode >= 200 && statusCode < 300) {
        println(s"✅ Successfully registered service at $registrationUrl")
      } else {
        println(s"❌ Failed to register service. Status: $statusCode, Body: $responseBody")
      }

    } catch {
      case ex: Exception =>
        println(s"💥 Exception in registerService: ${ex.getMessage}")
        ex.printStackTrace()
    }
  }



}
